### ISSUE TITLE
Test fixtures to generate valid objects for testing

### ISSUE THREAD

**Issue by yujuhong** [kubernetes/kubernetes#5121]
TL;DR, We need test fixtures that can generate valid objects for testing.

In #3854, we migrated the defaulting logic from the validation code to the decoder. As a result of that, many tests broke because they didn't expect the result object would be populated with default values.

These test follow such a pattern:
1. Define an "expected" object
2. Do the actual testing and produce an "actual" object
3. Compare the actual object with the expected object use `api.Semantic.DeepEqual()`

The expected object doesn't have all the fields filled, while the actual object would be populated with default values. In most cases, the test doesn't care about the default values. They only want to know the fields set in the expected object are the same as those in the actual object.

To avoid having to set every field (which has a default value) in the expect object, I added a `api.Semantic.DeepDerivative(x, y)` function. This function returns true if `y` is a derivative of `x` (note: the use of derivative is not very precise here). This means that if a field in `x` is set, the same field in `y` must be set to the same value. However, if a field in `x` is not set, we don't care what value is set in `y`.

The real solution to fix all this tests is to have some sort of test fixtures that can generate valid objects for testing. This would also clean up our test code quite a lot, and we can get rid of the `DeepDerivative` function.

/cc @bgrant0607, @lavalamp

**Comment by bgrant0607** [comment]
Text fixtures would likely be useful -- they'd be like "config generators" similar to kubectl run-container.

However, I think the first thing we should do is use the versioned representation of objects.

**Comment by roberthbailey** [comment]
Are we going to get to this for 1.2 or should we bump it to the next candidate milestone?

**Comment by bgrant0607** [comment]
Bump.

**Comment by kargakis** [comment]
@kubernetes/sig-testing-misc
